### PR TITLE
The loss function for each equation

### PR DESCRIPTION
**Pull request by KirillZubov**
https://github.com/SciML/NeuralPDE.jl/issues/320

### CODE REVIEW

**Comment by ChrisRackauckas** [comment]
Is this compatible with @zoemcc's changes? I want to make sure the direction is agreed upon here.

File: src/pinns_pde_solve.jl [hunk 1]
```diff
@@ -727,7 +727,7 @@ function get_loss_function(loss_function, lb,ub , strategy::QuadratureTraining;
             x = adapt(parameterless_type_θ,x)
             sum(abs2,loss_(x,θ), dims=2)
         end
-
+        # τ  = 1/lentgh(last_x)
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:730]:
```suggestion
        # τ  = 1/length(last_x)
```

File: src/pinns_pde_solve.jl [hunk 2]
```diff
@@ -800,7 +800,6 @@ function SciMLBase.discretize(pde_system::PDESystem, discretization::PhysicsInfo
     chain = discretization.chain
     initθ = discretization.init_params
     flat_initθ = if (typeof(chain) <: AbstractVector) vcat(initθ...) else  initθ end
-    # global typeofθ = typeof(Array(flat_initθ))
     global eltypeθ = eltype(flat_initθ)
     global parameterless_type_θ =  DiffEqBase.parameterless_type(flat_initθ)
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:804]: don't use globals for this. This is going to cause bugs.

File: src/pinns_pde_solve.jl [hunk 3]
```diff
@@ -42,19 +42,21 @@ struct PhysicsInformedNN{isinplace,C,T,P,PH,DER,PE,AL,K} <: AbstractPINN{isinpla
         else
             initθ = init_params
         end
+        flat_initθ = if (typeof(chain) <: AbstractVector) vcat(initθ...) else  initθ end
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:45]:
```suggestion
        flat_initθ = if (typeof(chain) <: AbstractVector) reduce(vcat,initθ) else  initθ end
```

Comment (ChrisRackauckas) [reply src/pinns_pde_solve.jl:45]: why is flattening necessary?

File: src/pinns_pde_solve.jl [hunk 4]
```diff
@@ -758,6 +760,9 @@ function SciMLBase.symbolic_discretize(pde_system::PDESystem, discretization::Ph
 
     chain = discretization.chain
     initθ = discretization.init_params
+    flat_initθ = if (typeof(chain) <: AbstractVector) vcat(initθ...) else  initθ end
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:763]: same here

File: src/pinns_pde_solve.jl (hunk 3 above)
Comment (KirillZubov) [reply src/pinns_pde_solve.jl]: in order to pass to `GalacticOptim.OptimizationProblem(f, flat_initθ)`

File: src/pinns_pde_solve.jl [hunk 5]
```diff
@@ -617,7 +617,7 @@ function get_numeric_derivative(parameterless_type_θ)
         end
 end
 
-Base.Broadcast.broadcasted(::typeof(get_numeric_derivative(AbstractVector)), phi,u,x,εs,order,θ) = get_numeric_derivative(AbstractVector)(phi,u,x,εs,order,θ)
+Base.Broadcast.broadcasted(::typeof(get_numeric_derivative(AbstractArray)), phi,u,x,εs,order,θ) = get_numeric_derivative(AbstractArray)(phi,u,x,εs,order,θ)
```
Comment (ChrisRackauckas) [inline src/pinns_pde_solve.jl:620]: why is this broadcast overload necessary?

Comment (KirillZubov) [reply src/pinns_pde_solve.jl:620]: in order to overload broadcasting for functions of `u` and `derivative`  in the internal representation
derivative.(some) ==  derivative(some) 
```
 Expr[:((cord, var"##θ#1508", phi, derivative, u, p)->begin
          begin
              let (θ,) = (cord[[1], :],)
                  (+).(derivative.(phi, u, cord, Vector{Float32}[[0.0049215667]], 1, var"##θ#1508"), u.(cord, var"##θ#1508", phi)) .- (cos).(θ)
              end
          end
      end)])
```

Comment (KirillZubov) [reply src/pinns_pde_solve.jl:620]: I would like to get rid of this, but for now, like this

File: test/NNPDE_tests.jl [hunk 6]
```diff
@@ -446,41 +427,26 @@ inn = 18
 chain = FastChain(FastDense(1,inn,Flux.σ),
                   FastDense(inn,inn,Flux.σ),
                   FastDense(inn,1))
+initθ = Float64.(DiffEqFlux.initial_params(chain))
```
Comment (KirillZubov) [inline test/NNPDE_tests.jl:430]: it looks like that BFGS stock with inner convert Float32/Float64 if all in Float64 is work fine

**Comment by KirillZubov** [comment]
different behavior for overload 'dottable' on different devices
```julia
julia> versioninfo()
Julia Version 1.6.0
Commit f9720dc2eb (2021-03-24 12:55 UTC)
Platform Info:
  OS: macOS (x86_64-apple-darwin19.6.0)
  CPU: Intel(R) Core(TM) i5-1038NG7 CPU @ 2.00GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, icelake-client)
Environment:
  JULIA_NUM_THREADS = 4

import Base.Broadcast
Base.Broadcast.dottable(x::Symbol) =  x !== :derivative
expr = :(derivative(phi, u, cord, Vector{Float32}[[0.0049215667]], 1, θ))
julia> bexpr = Base.Broadcast.__dot__(expr)
:(derivative(phi, u, cord, Vector{Float32}[[0.0049215667]], 1, θ))

julia> versioninfo()
Julia Version 1.6.0
Commit f9720dc2eb (2021-03-24 12:55 UTC)
Platform Info:
  OS: Linux (x86_64-pc-linux-gnu)
  CPU: Intel(R) Xeon(R) CPU @ 2.20GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-11.0.1 (ORCJIT, broadwell)

import Base.Broadcast
Base.Broadcast.dottable(x::Symbol) =  x !== :derivative
expr = :(derivative(phi, u, cord, Vector{Float32}[[0.0049215667]], 1, θ))
julia> bexpr = Base.Broadcast.__dot__(expr)
:(derivative.(phi, u, cord, Vector{Float32}[[0.0049215667]], 1, θ))
```

File: test/NNPDE_tests.jl (hunk 6 above)
Comment (KirillZubov) [reply test/NNPDE_tests.jl:432]: NN(x) + x 
convert(Float32 + Float64) -> Float64
and we get Float64 with Float32 precision

Comment (ChrisRackauckas) [reply test/NNPDE_tests.jl:432]: Yeah, it's up to the user to just make sure that types match.

**Comment by ChrisRackauckas** [comment]
>different behavior for overload 'dottable' on different devices

Don't overload that. That is massive type piracy.

**Comment by KirillZubov** [comment]
@ChrisRackauckas  I think it is done. could you please review it

File: docs/src/pinn/debugging.md [hunk 7]
```diff
@@ -52,9 +54,9 @@ dim = length(domains)
 dx = 0.1
 strategy = NeuralPDE.GridTraining(dx)
 
-_pde_loss_function = NeuralPDE.build_loss_function(eq,indvars,depvars,phi, derivative,chain,initθ,strategy)
+_pde_loss_function = NeuralPDE.build_loss_function(eq,indvars,depvars,phi, derivative,chain,initθ,eltypeθ,strategy)
```
Comment (ChrisRackauckas) [inline docs/src/pinn/debugging.md:57]: could it be less breaking to just `eltype(inittheta)` inside of the function?

Comment (KirillZubov) [reply docs/src/pinn/debugging.md:57]: yep, it makes sense. also we can pass only `chain` because initθ = DiffEqFlux.initial_params(chain)

**Pull request merged**